src/main/java/ca/wolfram/beta/symath/parser/AntlrExpressionListener.java

```diff
@@ -0,0 +1,316 @@
+import ca.wolfram.beta.symath.parser.ExpressionParser.ScientificContext;
+import ca.wolfram.beta.symath.parser.ExpressionParser.TermContext;
+import ca.wolfram.beta.symath.parser.ExpressionParser.VariableContext;
+import ca.wolfram.beta.symath.parser.ExpressionParser.WholenumberContext;
+
+public class AntlrExpressionListener implements ExpressionListener{
+
+    private Stack<ArrayList<MathNode>> nodesStack = new Stack<ArrayList<MathNode>>();
+    private MathNode rootNode;
+
+    public MathNode getExpression() {
+        return rootNode;
+    }
+
+    // Auto generated by the parser
+
+    @Override
+    public void enterEveryRule(ParserRuleContext arg0) {
+        // TODO Auto-generated method stub
+    }
+
+    @Override
+    public void exitEveryRule(ParserRuleContext arg0) {
+        // TODO Auto-generated method stub
+    }
+
+    @Override
+    public void visitErrorNode(ErrorNode arg0) {
+        // TODO Auto-generated method stub
+    }
+
+    @Override
+    public void visitTerminal(TerminalNode arg0) {
+        // TODO Auto-generated method stub
+    }
+
+    @Override
+    public void enterProg(ProgContext ctx) {
+        //System.out.println("Enter Prog: " + ctx.getText());
+        nodesStack.push(new ArrayList<MathNode>());
+    }
+
+    @Override
+    public void exitProg(ProgContext ctx) {
+        rootNode = nodesStack.pop().get(0);
+    }
+
+    // Unused
+
+    @Override
+    public void enterEquation(EquationContext ctx) {
+        // TODO Auto-generated method stub
+    }
+
+    @Override
+    public void exitEquation(EquationContext ctx) {
+        // TODO Auto-generated method stub
+    }
+
+    // for an expression in the format "a+b+c-d-e...". Created in the format SubtractNode(AddNode(a, b, c), d, e)
+
+    @Override
+    public void enterExpression(ExpressionContext ctx) {
+        nodesStack.push(new ArrayList<MathNode>());
+        //System.out.println("Entered Expression " + ctx.getText() + ", current stack size: " + nodesStack.size());
+    }
+
+    @Override
+    public void exitExpression(ExpressionContext ctx) {
+        ArrayList<MathNode> children = nodesStack.pop();
+
+        ArrayList<MathNode> add = new ArrayList<MathNode>();
+        ArrayList<MathNode> subtract = new ArrayList<MathNode>();
+        add.add(children.get(0));
+        for (int i = 1; i < children.size(); i++) {
+            if( ctx.getChild(i*2-1).getText().equals("+")) {
+                add.add(children.get(i));
+            } else {
+                subtract.add(children.get(i));
+            }
+        }
+
+        MathNode expression;
+        if (add.size() == 1) expression = add.get(0);
+        else expression = AddNode.create(add);
+        if (!subtract.isEmpty()) {
+            subtract.add(0, expression);
+            expression = SubtractNode.create(subtract);
+        }
+
+        nodesStack.peek().add(expression);
+        //System.out.println("Exit Expression, current stack size: " + nodesStack.size());
```

Comment:
to help with debugging, I usually have the following

```
static boolean debug = false;

void print(String s, Object... o) {
    if (debug) System.out.println(String.format(s, o));
}
```

In case we need the print statements again.